### DISCUSSION TITLE
もしよければ、シーケンスについてお願いします

### DISCUSSION THREAD

**Discussion by ki052020** [fffsharp/Forum#3] [質問]
F# の超初心者です！すみません、時間のあるときにコメントをもらえたら、とても嬉しいです。
良くない例として以下のコードを見つけたのですが、以下の例を見るとシーケンスの使い所ってどんなとこなんだろう？と考えています。
(A) をコメントアウトして、(B) の方を有効にすれば、なるほどなー、というだけだったのだけど、
length を評価するだけで内容が破棄されるのがシーケンスなのか、と思うとモヤモヤします。
F# のまともなコードを書いてないからだと思うのですが、すみませんが何か指針をもらえたら嬉しいです！

-----------
```
let OnSeq (d0) =
    let mutable c = 0
    let d1 = 
        Seq.map 
            (fun (s: string) -> 
                c <- c + 1
                s + " d1 " + string c
            ) 
            d0

    if Seq.length d1 > 2 then  // (A)
    //if true then  // (B)
        let d2 = 
            Seq.map
                (fun (s: string) -> 
                    c <- c + 1
                    s + " d2 " + string c
                ) 
                d1
        Seq.toList d2
    else
        []

[<EntryPoint>]
let main argv =
    let lst = [ "abc"; "def"; "ghi" ]
    let result = OnSeq(lst)
    let _ = List.iter (fun (s: string) -> stdout.WriteLine(s)) result
    0
```
-----------
実行結果
abc d1 4 d2 5
def d1 6 d2 7
ghi d1 8 d2 9

**Comment by tatsuya-midorikawa** [comment]
@ki052020 さん

`seq<'T>` は評価されるたびに処理が実行されます。  
逆に言えば、評価がされなければ、処理は実行されません。  
今回のサンプル コードの場合、`Seq.length` と `Seq.toList` のところで `d1` が評価されているため、`d1` の処理が 2 回実行され、例示いただいている結果になっています。  

言葉だけではわかりにくいと思いますので、簡単なサンプルを用意してみます。  
例えば、`List<'T>` を操作するモジュールとして `List モジュール` というものがあります。  
これを、`Seq モジュール` を使った場合と比較してみます。  

```fsharp
let xs = [ "abc"; "def"; "ghi" ]
let ys = xs |> List.map (fun x ->    // List モジュールを利用
  printf $"%s{x} "
  x.ToUpper())

printfn "\n---"

for y in ys do
  printfn $"%s{y}"
```  
この場合、どういった出力が期待されるでしょうか？   
処理の実行順を考えると、以下のようなものを期待するのが普通だと思います。  
```
abc def ghi
---
ABC
DEF
GHI
```  
では、実際に実行するとどうなるかというと、期待通りの結果になります。  
これは `List モジュール` が実行を先送りにせず、すぐさま処理を行うからです。  
![image](https://user-images.githubusercontent.com/78302178/153961290-85005661-5823-4df5-9509-a8d6f0280618.png)  
<br>
次に、`Seq モジュール` に置き換えてみます。  

```fsharp
let xs = [ "abc"; "def"; "ghi" ]
let ys = xs |> Seq.map (fun x ->    // Seq に変更
  printf $"%s{x} "
  x.ToUpper())

printfn "\n---"

for y in ys do
  printfn $"%s{y}"
```  
果たしてこの場合はどのように実行されるのでしょうか？  
実際に実行してみると、以下のようになります。  
![image](https://user-images.githubusercontent.com/78302178/153961587-bbd42162-17b5-47ba-9c81-b872c5e38b5e.png)  

`printfn "\n---"` の後に、`Seq.map` の中身が実行されていることがわかります。  
また、`for y in ys do` の箇所で値を 1 つ取得するたびに、処理が実行されていることも見て取れます。  
つまり、実際に使うタイミングになって、初めて `Seq.map` の中身の処理が実行されたわけです。  
<br>  
このように、処理の実行を先送りにしたい場合に `Seq` を使うと効果的な場合があります。  
例えば、いろいろな条件をクリアした場合にのみ処理を行いたい場合や、特定の条件の場合にのみ処理を追加したい場合などに利用したりする場合などです。  
ただし、状況によって効果的な場合とそうでない場合がありますので、上記のような状況だからといって、即 「こういう場合は Seq で処理を行う」とはしない方が賢明だと思います。  
<br>

雰囲気だけですが、なんとなく伝わりましたでしょうか？

**Reply by tatsuya-midorikawa** [reply]
■ 追記  
Seq について補足しておきます。  
上記で、
> `seq<'T>` は評価されるたびに処理が実行されます。

と書きましたが、ちょっとわかりにくいと思ったので、簡単なサンプルを作りました。
<br>
`List モジュール` のサンプルから見ていきます。  
```fsharp
let xs = [ "abc"; "def"; "ghi" ]
let ys = xs |> List.map (fun x ->    // List モジュールを使用
  printf $"%s{x} "
  x.ToUpper())

printfn "\n---"

// 1 回目
for y in ys do
  printfn $"%s{y}"
// 2 回目
for y in ys do
  printfn $"%s{y}"
```  
この実行結果は、以下のようになります。  
![image](https://user-images.githubusercontent.com/78302178/153963964-00c69e9a-db69-4411-ae7e-d50487365f04.png)  

`List.map` の中身の処理が、1 度しか行われていないことがわかります。  
では、`Seq モジュール` の場合、どうなるのでしょうか？  

```fsharp
let xs = [ "abc"; "def"; "ghi" ]
let ys = xs |> Seq.map (fun x ->    // Seq に変更
  printf $"%s{x} "
  x.ToUpper())

printfn "\n---"

// 1 回目
for y in ys do
  printfn $"%s{y}"
// 2 回目
for y in ys do
  printfn $"%s{y}"
```  
これの実行結果は、以下のようになります。  
![image](https://user-images.githubusercontent.com/78302178/153964132-63597f51-c6ad-4de7-a4dc-907a23b364d3.png)  

2 回目にも小文字の文字列が出力されていることから、`Seq.map` の処理が毎回呼ばれていることがわかります。  
そのため、今回でいうところの `ys` を何度もループさせて利用させることがわかっている場合には、一度 `toList` や `toArray` をして、評価を確定したあとの値を利用する方が良いです。  
この点に注意しないと、何度も処理が実行されてしまい、パフォーマンスの低下につながってしまいます。  
だからといって、`toList` や `toArray` をむやみやたらに呼び出すのも同様にパフォーマンスの低下につながるので、注意が必要です。

**Reply by ki052020** [reply]
とても丁寧に答えてくださって、本当にありがとうございます！
どういうところに注意すべきか、というのはサンプルコードからよく伝わりとても参考になります。
F# をよく知っている方のサンプルコードは、参考になるところが多く大変助かります！
遅延評価をするものを簡単に実現できそうなんだな、という雰囲気はよく分かりました。いろいろ試しに書いてみて、もう少し調べてみようと思います。
忙しい中、スクショまで付けて返答を作ってくださって、すごく感謝しています。ありがとうございました。

**Reply by sayurin** [reply]
ところで

> ```
>     let d1 = 
>         Seq.map 
>             (fun (s: string) -> 
>                 c <- c + 1
>                 s + " d1 " + string c
>             ) 
>             d0
> ```

と

> ```
>    //if true then  // (B)
>         let d2 = 
>             Seq.map
>                 (fun (s: string) -> 
>                     c <- c + 1
>                     s + " d2 " + string c
>                 ) 
>                 d1
> ```

とで `let mutable c = 0` を共用しているのは意図した設計なのでしょうか？
「`d2` の開始時点で `c` に何の値が格納されていても構わない」と読み取れますが、一般的にはあり得ない設計です。
「学習目的で特に意図はなかった」なのかもしれませんが、あるべき姿を想像してみてください。

このような指摘を行うのには理由があり、 F# では不変性を重視しています。`mutable`を付けなければ変数の値を変更できないのもそのためです。値が変化しなければ、いつ参照しても同じ値が得られプログラム的に安全なわけです。逆に値が変化する場合、適切なタイミングで参照しなければプログラムとして不正になります。
今回の質問も `c` が変化し、それを参照するタイミングが違うことに起因する問題です。

----

なお [`Seq.map`](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-collections-seqmodule.html#map) を利用されていますが、 [`Seq.mapi`](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-collections-seqmodule.html#mapi) も用意されています。こちらはインデックス付きなため、`c`のようなカウンタ変数を用意せずに済みます。
ゼロオリジンなので `+ 1`しています。

```
let d1 = Seq.mapi (fun c s -> s + " d1 " + string (c + 1)) d0
```

あと、 F# 5 では[文字列補間](https://docs.microsoft.com/ja-jp/dotnet/fsharp/whats-new/fsharp-50#string-interpolation)が導入されました。文字列を構築する際、より直感的で解釈しやすい記述ができます。

```
let d1 = Seq.mapi (fun c s -> $"{s} d1 {c + 1}") d0
```

**Reply by ki052020** [reply]
いろいろな意見が聞けて嬉しいです！
この例は「良くない例」として挙げられていたもので、元の作者の人もこういうところで c を使うのは良くないんだよ、ということは一言添えてありました。当然「d2 の開始時点で c に何の値が格納されていても構わない」という設計です(^^;。つまり、c の値ってインデックスとかを表しているんじゃないんですよ。単に実行順序の確認のヒントとして c を使ったみたいです。
F# は超初心者ですが、関数型言語で副作用を持つものを書くのは避けたほうがいい、というのは大事ですよね。また他に気付いたことなどがあれば指摘してもらえると知識が増えて嬉しいです！
